Migrate to PEG parser. Introduce boolean operators and constants.

stephenquan opened the pull request:
<!--
 Hello, and thank you for your interest in contributing to the .NET MAUI Toolkit! 

 Before you submit please check that this work relates to one of the following:
 - Bug fix
    If you haven't yet opened an Issue that reports the bug in detail, provides a reproduction sample, and has been verified + reproduced by a member of the .NET MAUI Toolkit core team, please do that before submitting a Pull Request.
 - Feature/Proposal
    If you haven't yet submitted a Proposal that has been Championed by a .NET MAUI core team member, please instead open a Discussion at https://github.com/communitytoolkit/maui/discussions/new where we can discuss the pros/cons of the feature and its implementation. 
 Any PR submitted that does not fit with the above options will be closed.
 -->

 ### Description of Change ###

 - MultiMathExpressionConverterPage.xaml
   - Add VerticalStackLayout and move existing sample inside
   - Added a Label with a DataTrigger on "x0 + x1 + x2 >= 60" to demonstrate the comparison operator
 - MathOperator.shared.cs
   - Replaced `double` with `object`
   - MathOperatorPrecedence obsolete (part of the original parsing engine but no longer relevant in new engine)
 - MathExpressionConverterTests.cs
   - New tests for comparison, equality, and boolean operators
   - Certain operations allow nulls now
 - MathExpression.shared.cs
   - Refactor/rewrite to use Parsing expression grammar
   - The grammar parity with original operators `+`, `-`, `*`, `/` and `^` as well as math functions
   - Added to grammar `and`, `or`, `?`, `:`, `>=`, `>`, `<=`, `<`, `==`, `!=` as well as `true` and `false` constants
   - Relax null strictness (e.g. equality and ternary operators can have nulls)
 - MathExpressionConverter.shared.cs
   - Replaced `double` with `object`
 - MultiMathExpressionConverter.cs
   - Relax null strictness (e.g. equality and ternary operators can have nulls)

General comments:
 - Original implementation sanitizes all inputs by calling double.Parse(value.ToString) rejecting nulls
 - New implementation calls Convert.ToDouble() when needed and, in certain situations, allows nulls
 - When a boolean value is needed, boolean coercion uses the following rules:
   - If value is null, return false
   - If value is a boolean, return value
   - If value is a double, return true if value != 0 && value != NaN, false otherwise
   - If value is a string, return true, if value is not null or empty, false otherwise
 
 ### Linked Issues ###

https://github.com/CommunityToolkit/Maui/issues/2181

 ### PR Checklist ###

 - [x] Has a linked Issue, and the Issue has been `approved`(bug) or `Championed` (feature/proposal)
 - [x] Has tests (if omitted, state reason in description)
 - [x] Has samples (if omitted, state reason in description)
 - [x] Rebased on top of `main` at time of PR
 - [x] Changes adhere to [coding standard](https://github.com/CommunityToolkit/Maui/blob/main/CONTRIBUTING.md#contributing-code---best-practices)
 - [x] Documentation created or updated: https://github.com/MicrosoftDocs/CommunityToolkit/pull/465

 ### Additional information ###

brminnick commented on src/CommunityToolkit.Maui.UnitTests/Converters/MathExpressionConverterTests.cs:
```diff
@@ -85,9 +124,9 @@ public void MathExpressionConverterNullInputTest()
 		Assert.Throws<ArgumentNullException>(() => ((ICommunityToolkitValueConverter)new MathExpressionConverter()).ConvertBack(0.0, null, null, null));
 #pragma warning restore CS8625 // Cannot convert null literal to non-nullable reference type.
-		Assert.Throws<ArgumentNullException>(() => ((ICommunityToolkitValueConverter)new MathExpressionConverter()).Convert(null, typeof(bool), "x", null));
+		//Assert.Throws<ArgumentNullException>(() => ((ICommunityToolkitValueConverter)new MathExpressionConverter()).Convert(null, typeof(bool), "x", null));
```
Is it possible to amend this Unit Test so that it passes? Or is this Assert failing because of a bug in the new code?

It's not a best-practice to have commented-out assert statements in the Unit Test project.

stephenquan replied:
@brminnick the assertion was failing in the new code. I didn't understand the assertion. I will try and figure it out.

stephenquan replied:
@brminnick I understand the regression is actually by design by the PR. All values passed thru to the MathExpressionConverters was being sanitized with this logic which raises an exception if any of the inputs are null:

```c#
var valueString = value?.ToString() ?? throw new ArgumentException("Values cannot be null.");
var xValue = double.Parse(valueString);
```

In the new code, I allow nulls, particularly, in the conditional `conditonal ? true_value : false_value` both values on the right can be nullable. In the unit test, it was asking if we past `x=null` to an math expression should it raise an exception. The new behavior it returns null without raising an exception. I believe this to be acceptable improvement, hence, I will rewrite the test to reflect.

stephenquan replied:
@brminnick I have amended this unit test

brminnick commented on src/CommunityToolkit.Maui.UnitTests/Converters/MathExpressionConverterTests.cs:
```diff
@@ -85,9 +124,9 @@ public void MathExpressionConverterNullInputTest()
+		//Assert.Throws<ArgumentNullException>(() => ((ICommunityToolkitValueConverter)new MathExpressionConverter()).Convert(null, typeof(bool), "x", null));
 		Assert.Throws<ArgumentNullException>(() => ((ICommunityToolkitValueConverter)new MathExpressionConverter()).Convert(null, typeof(bool), null, null));
-		Assert.Throws<ArgumentNullException>(() => ((ICommunityToolkitValueConverter)new MathExpressionConverter()).ConvertBack(null, typeof(bool), null, null));
+		//Assert.Throws<ArgumentNullException>(() => ((ICommunityToolkitValueConverter)new MathExpressionConverter()).ConvertBack(null, typeof(bool), null, null));
```
Is it possible to amend this Unit Test so that it passes? Or is this Assert failing because of a bug in the new code?

It's not a best-practice to have commented-out assert statements in the Unit Test project.

stephenquan replied:
I have amended this unit test.

stephenquan commented:
@dotnet-policy-service agree company="Esri"

stephenquan commented:
Hi @brminnick since your review, I have:
 - attempted to update the fork with the base branch
 - corrected the unit test issues you raised
 - added more unit test for better coverage of the boolean operators
 - updated the handling of nulls

The fork is ready for another round of review

pictos commented on src/CommunityToolkit.Maui/Converters/MathExpressionConverter/MathExpression.shared.cs:
```diff
@@ -4,94 +4,179 @@
+	internal static bool __bool(object? value)
+	{
+		if (value is null)
+		{
```
would be great if we can make this use switch expression:

```cs
bool Bla(object b) =>
    b switch 
    {
        bool x => x,
        null => false,
        double doubleValue => doubleValue != 0 && doubleValue != double.NaN,
        string stringValue => !string.IsNullOrEmpty(stringValue),
        _ => Convert.ToBoolean(b)         
    };
    
```

stephenquan replied:
Thanks @pictos. I migrated to the new C# switch statement as per your feedback.

pictos commented on src/CommunityToolkit.Maui/Converters/MathExpressionConverter/MathExpression.shared.cs:
```diff
@@ -4,94 +4,179 @@
+	Operator,
+};
+
+class MathToken
+{
+	public MathToken(MathTokenType type, string text, object? value)
+	{
+		Type = type;
+		Text = text;
+		Value = value;
+	}
+
+	public MathTokenType Type { get; }
+	public string Text { get; }
+	public object? Value { get; }
+};
```
Let's use a record, this will give us the Equals and getHash implemented by value. Which is, probably, what we want here.

```suggestion
record MathToken (MathTokenType type, string text, object? value);
```

stephenquan replied:
Thanks @pictos. I implemented recorded as per your feedback.

pictos commented:
@stephenquan here's the build log:

```
D:\a\1\s\src\CommunityToolkit.Maui\Converters\MathExpressionConverter\MathExpressionConverter.shared.cs(25,10): error CS8603: Possible null reference return. [D:\a\1\s\src\CommunityToolkit.Maui\CommunityToolkit.Maui.csproj::TargetFramework=net8.0-ios]
D:\a\1\s\src\CommunityToolkit.Maui\Converters\MathExpressionConverter\MathExpressionConverter.shared.cs(25,10): error CS8603: Possible null reference return. [D:\a\1\s\src\CommunityToolkit.Maui\CommunityToolkit.Maui.csproj::TargetFramework=net8.0-windows10.0.19041.0]
D:\a\1\s\src\CommunityToolkit.Maui\Converters\MathExpressionConverter\MathExpressionConverter.shared.cs(25,10): error CS8603: Possible null reference return. [D:\a\1\s\src\CommunityToolkit.Maui\CommunityToolkit.Maui.csproj::TargetFramework=net8.0-tizen]
D:\a\1\s\src\CommunityToolkit.Maui\Converters\MathExpressionConverter\MathExpressionConverter.shared.cs(25,10): error CS8603: Possible null reference return. [D:\a\1\s\src\CommunityToolkit.Maui\CommunityToolkit.Maui.csproj::TargetFramework=net8.0-maccatalyst]
D:\a\1\s\src\CommunityToolkit.Maui\Converters\MathExpressionConverter\MathExpressionConverter.shared.cs(25,10): error CS8603: Possible null reference return. [D:\a\1\s\src\CommunityToolkit.Maui\CommunityToolkit.Maui.csproj::TargetFramework=net8.0-android]
  
```

stephenquan commented:
Thanks @pictos, I have reviewed the CS8063 possible null reference return issues and have resolved it by making nullability updates to both the MathExpression.shared.cs and MathExpressionConverter.shared.cs. The nullability updates also required additional Asserts added to the MathExpressionConverterTests.cs. With these changes, the build checks are now passing.

pictos commented on src/CommunityToolkit.Maui/Converters/MathExpressionConverter/MathExpression.shared.cs:
```diff
@@ -4,94 +4,144 @@
 	}
 
-	internal string Expression { get; }
+	internal string expr { get; }
+
+	internal int exprIndex { get; set; } = 0;
+
+	internal Match patternMatch { get; set; } = Match.Empty;
+
+	internal List<MathToken> rpn { get; } = new();
```
For those can you follow the C# property name pattern?

stephenquan replied:
Hi @pictos, I would like to follow the correct C# property name pattern. I am assuming this to be necessary capitalization, e.g. `Expr`, `ExprIndex`, `PatternMatch`, `RPN`? Or did you have something else in mind?

bijington replied:
I would assume so. I would personally prefer full words over abbreviations e.g. ExpressionIndex over ExprIndex

pictos commented on src/CommunityToolkit.Maui/Converters/MathExpressionConverter/MathExpression.shared.cs:
```diff
@@ -4,94 +4,144 @@
-		var rpn = GetReversePolishNotation(Expression);
+		if (!ParseExpression())
+		{
+			throw new ArgumentException("Invalid math expression.");
```
if we throw here, will the app break? If so, I think that we shouldn't throw an exception, but log something for the user or having some event or property that will tell when an invalid expression is passed.

stephenquan replied:
Hi @pictos, the exceptions were there because I was trying to be consistent with the original implementation, however, the exceptions limit the effectiveness of hot reload, so, as per your request I am more than willing to remove them. I am open to suggestions on how to inform the developer perhaps `Trace.TraceWarning`?

pictos replied:
Yeah, TraceWarning is good for me. In the future we can expand it , if devs request

pictos reviewed (commented):
I think that we're getting really close to merge this!

pictos commented on src/CommunityToolkit.Maui/Converters/MathExpressionConverter/MultiMathExpressionConverter.shared.cs:
```diff
@@ -28,21 +28,7 @@ public class MultiMathExpressionConverter : MultiValueConverterExtension, ICommu
-		}
-
-		var math = new MathExpression(expression, args);
+		var math = new MathExpression(expression, values!);
```
please don't use the `!` operator

brminnick replied:
Agreed. Specifically, do not use the Null Forgiving Operator.

stephenquan commented:
@pictos @brminnick as requested, I have removed the Null Forgiving Operator from the MultiMathExpressionConverter. I have corrected the null checks in the MultiMathExpressionConverter and that resulted in a necessary minor update to the unit tests.

stephenquan commented:
Hi all, we are copies of these classes in our project with great success in our projects. I am eager to help out and remove any blockers on this PR.

pictos reviewed (dismissed):
LGTM

stephenquan commented:
Hi guys, I think this PR is good to merge.

bijington commented:
> Hi guys, I think this PR is good to merge.

Thanks. We are trying to get to the point of merging #2215 before any other PRs due to that size of it and that it provides .NET 9.0 support. Then we would ideally merge PRs like this. @brminnick is that correct order for our plan?

stephenquan commented:
I have some new developers working with Boolean operators in XAML, but they struggled with the syntax.

To simplify I things added new commits which adds XAML-friendly alternatives. Additionally, I added more coverage to the logical AND/OR unit tests.

Unfortunately, in doing so, 3 of the 6 automated test are failing and I am not sure why. Also, the PR will now require a re-review.

| Operator | XAML         | Alternate |
| :------: | :----------: | :-------: |
| `&&`     | `&amp;&amp;` | and       |
| `\|\|`   | `\|\|`       | or        |
| `>=`     | `&gt;=`      | ge        |
| `>`      | `&gt;`       | gt        |
| `<=`     | `&lt;=`      | le        |
| `<`      | `&lt;`       | lt        |

Here are some examples of the alternate syntax:

| Original                          | Alternate            |
| :-------------------------------: | :------------------: |
| `x &gt;= 0 ? x1 : x2`             | `x ge 0 ? x1 : x2`   |
| `x &amp;&amp; !x1`                | `x and !x1`          |
| `x &gt;= 0 &amp;&amp; x &lt;= 10` | `x ge 0 and x le 10` |

stephenquan commented:
The azure pipelines https://dev.azure.com/dotnet/CommunityToolkit/_build/results?buildId=112270&view=logs&j=67da6a11-18ef-5f70-9b43-fae7e6c83e18&t=54ba8226-26d8-53c8-39e8-392f99f22929&l=538 is showing 11 new errors that I do not know how to resolve. These 11 new errors do not relate the commits done, so, I'm hoping they will automatically resolve when the fork catches up with the main branch:

D:\a\1\s\samples\CommunityToolkit.Maui.Sample\MauiProgram.cs(38,20): error CS0234: The type or namespace name 'Composition' does not exist in the namespace 'Microsoft.UI' (are you missing an assembly reference?) [D:\a\1\s\samples\CommunityToolkit.Maui.Sample\CommunityToolkit.Maui.Sample.csproj::TargetFramework=net8.0-windows10.0.19041.0]
D:\a\1\s\samples\CommunityToolkit.Maui.Sample\MauiProgram.cs(39,20): error CS0234: The type or namespace name 'Windowing' does not exist in the namespace 'Microsoft.UI' (are you missing an assembly reference?) [D:\a\1\s\samples\CommunityToolkit.Maui.Sample\CommunityToolkit.Maui.Sample.csproj::TargetFramework=net8.0-windows10.0.19041.0]
D:\a\1\s\samples\CommunityToolkit.Maui.Sample\MauiProgram.cs(40,25): error CS0234: The type or namespace name 'Media' does not exist in the namespace 'Microsoft.UI.Xaml' (are you missing an assembly reference?) [D:\a\1\s\samples\CommunityToolkit.Maui.Sample\CommunityToolkit.Maui.Sample.csproj::TargetFramework=net8.0-windows10.0.19041.0]
D:\a\1\s\samples\CommunityToolkit.Maui.Sample\Platforms\Windows\App.xaml.cs(6,28): error CS0246: The type or namespace name 'MauiWinUIApplication' could not be found (are you missing a using directive or an assembly reference?) [D:\a\1\s\samples\CommunityToolkit.Maui.Sample\CommunityToolkit.Maui.Sample.csproj::TargetFramework=net8.0-windows10.0.19041.0]
D:\a\1\s\samples\CommunityToolkit.Maui.Sample\Platforms\MacCatalyst\AppDelegate.cs(6,28): error CS0246: The type or namespace name 'MauiUIApplicationDelegate' could not be found (are you missing a using directive or an assembly reference?) [D:\a\1\s\samples\CommunityToolkit.Maui.Sample\CommunityToolkit.Maui.Sample.csproj::TargetFramework=net8.0-maccatalyst]
D:\a\1\s\samples\CommunityToolkit.Maui.Sample\Platforms\iOS\AppDelegate.cs(6,28): error CS0246: The type or namespace name 'MauiUIApplicationDelegate' could not be found (are you missing a using directive or an assembly reference?) [D:\a\1\s\samples\CommunityToolkit.Maui.Sample\CommunityToolkit.Maui.Sample.csproj::TargetFramework=net8.0-ios]
D:\a\1\s\samples\CommunityToolkit.Maui.Sample\obj\Release\net8.0-android\lp\148\jl\res\layout\tabbar.xml(2): error APT2260: attribute tabGravity (aka com.microsoft.CommunityToolkit.Maui.Sample:tabGravity) not found. [D:\a\1\s\samples\CommunityToolkit.Maui.Sample\CommunityToolkit.Maui.Sample.csproj::TargetFramework=net8.0-android]
D:\a\1\s\samples\CommunityToolkit.Maui.Sample\obj\Release\net8.0-android\lp\148\jl\res\layout\tabbar.xml(2): error APT2260:  [D:\a\1\s\samples\CommunityToolkit.Maui.Sample\CommunityToolkit.Maui.Sample.csproj::TargetFramework=net8.0-android]
D:\a\1\s\samples\CommunityToolkit.Maui.Sample\obj\Release\net8.0-android\lp\148\jl\res\layout\tabbar.xml(2): error APT2260: attribute tabIndicatorColor (aka com.microsoft.CommunityToolkit.Maui.Sample:tabIndicatorColor) not found. [D:\a\1\s\samples\CommunityToolkit.Maui.Sample\CommunityToolkit.Maui.Sample.csproj::TargetFramework=net8.0-android]
D:\a\1\s\samples\CommunityToolkit.Maui.Sample\obj\Release\net8.0-android\lp\148\jl\res\layout\tabbar.xml(2): error APT2260:  [D:\a\1\s\samples\CommunityToolkit.Maui.Sample\CommunityToolkit.Maui.Sample.csproj::TargetFramework=net8.0-android]
D:\a\1\s\samples\CommunityToolkit.Maui.Sample\obj\Release\net8.0-android\lp\148\jl\res\layout\tabbar.xml(2): error APT2260: attribute tabMode (aka com.microsoft.CommunityToolkit.Maui.Sample:tabMode) not found. [D:\a\1\s\samples\CommunityToolkit.Maui.Sample\CommunityToolkit.Maui.Sample.csproj::TargetFramework=net8.0-android]
D:\a\1\s\samples\CommunityToolkit.Maui.Sample\obj\Release\net8.0-android\lp\148\jl\res\layout\tabbar.xml(2): error APT2260:  [D:\a\1\s\samples\CommunityToolkit.Maui.Sample\CommunityToolkit.Maui.Sample.csproj::TargetFramework=net8.0-android]
C:\hostedtoolcache\windows\dotnet\packs\Microsoft.Android.Sdk.Windows\34.0.145\tools\Xamarin.Android.Aapt2.targets(156,3): error APT2061: failed linking file resources. [D:\a\1\s\samples\CommunityToolkit.Maui.Sample\CommunityToolkit.Maui.Sample.csproj::TargetFramework=net8.0-android]
C:\hostedtoolcache\windows\dotnet\packs\Microsoft.Android.Sdk.Windows\34.0.145\tools\Xamarin.Android.Aapt2.targets(156,3): error APT2061:  [D:\a\1\s\samples\CommunityToolkit.Maui.Sample\CommunityToolkit.Maui.Sample.csproj::TargetFramework=net8.0-android]

brminnick commented:
Hi @stephenquan! Could you update this PR to resolve the merge conflicts? 

We've finally merged the massive .NET 9 PR today and it introduced a few merge conflicts across our open PRs.

brminnick commented on src/CommunityToolkit.Maui.UnitTests/Converters/MathExpressionConverterTests.cs:
```diff
@@ -29,34 +29,203 @@ public void MathExpressionConverter_ReturnsCorrectResult(string expression, doub
+	[InlineData("Cat", false, false, "Cat")]
+	[InlineData("", false, "", false)]
+	[InlineData(false, "", false, "")]
+	[InlineData(null, "Cat", null, "Cat")]
```
Hey @stephenquan! There's one thing I'm having trouble wrapping my head around in this PR: How/why we are evaluating `null` using boolean operators, as demonstrated here in this test? 

If I'm understanding this code correctly, the below line of code is equivalent to evaluating the boolean math equation of `null OR null`, right?

```cs
mathExpressionConverter.Convert(variables, typeof(object), "x || x1")
```

I haven't studied boolean algebra since college, but, my understanding is that boolean math does not include the concept of `null`. Is my understanding correct?

brminnick replied:
My gut tells me that we should be throwing an `Exception` in the scenarios where `null` is being used in a Boolean Algebraic equation.

In other words, `MathExpression.Calculate()` should not return `null` in these scenarios and instead throw an exception alerting the developer that they've input an invalid equation.

stephenquan replied:
@brminnick thanks for your question. The inspiration for logical-AND and logical-OR operators is derived from both Javascript and Python with respect to how they handle truthy and falsy concepts. For reference:

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Logical_OR
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Logical_AND

And the unit test mirror the results from the Javascript documentation examples.

The Python reference is under the section https://docs.python.org/3/library/stdtypes.html

If I wanted to implement the following case statement

```
// states of water
if (x0 < 0) then x1 // "ice";
else if (x0 < 100) then x2 // "water"
else x3 // "air"
```

I could write it as:

```
// using the conditional operator
x0 < 0 ? x1
       : x0 < 100
         ? x2
         : x3
```

Or, using the truthy and falsy states, I could rewrite it as

```
// using the truthy and falsy application of logical and and logical or
     x0 < 0 and x1
  or x0 < 100 and x2
  or x3
```

stephenquan replied:
Javascript 

```javascript
console.log(null || null); // output: null
console.log(null && null); // output: null
```

Python

```python
>>> print(None or None)
None
>>> print(None and None)
None
```

bijington replied:
It's the equivalent of a `Nullable<bool>` expression so I think it's fine

stephenquan replied:
@brminnick, I hope my explanation is satisfactory. Shall we proceed and mark this conversation as resolved?

brminnick reviewed (approved):
Thanks @stephenquan! Approved ✅
